### PR TITLE
Accept ::class as literal string

### LINKED ISSUE #332: class-string vs literal-string in QueryBuilder
With the [recent addition of literal-types to QueryBuilder](https://github.com/phpstan/phpstan-doctrine/pull/327/files#diff-197b8a38b2ae878d60d27968735dd0b7208e015f401be1b2c5761add8fe8b85dR61) (https://github.com/phpstan/phpstan-doctrine/pull/327), I'm getting warnings when passing `class-string` there: `Parameter #1 $from of method Doctrine\ORM\QueryBuilder::from() expects literal-string, class-string given`

This is the simplified example to reproduce the issue:

```php
    /**
     * @param class-string $className
     */
    public function getSortable(
        string $className,
    ): ?Sortable
    {
        return $this
            ->entityManager
            ->createQueryBuilder()
            ->select('sortable')
            ->from($className, 'sortable')
            ->getQuery()
            ->getOneOrNullResult();
    }
```

What is do you think is the correct way to handle this? To me it seems that `class-string` is more specific type than the `literal-string`.

cc @VincentLanglet

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/7823
closes https://github.com/phpstan/phpstan-doctrine/issues/332

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
I disagree with this fix. class-string isn't always literal string.
Rather than that, `$x::class` should be an intersection with literal-string. You can implement that in MutatingScope::resolveType() and also verify that in NodeScopeResolverTest.

**Comment by VincentLanglet** [comment]
> I disagree with this fix. class-string isn't always literal string.

@craigfrancis didn't give an example of class-string which are not literal-string in https://github.com/phpstan/phpstan-doctrine/issues/332#issuecomment-1134878945

But I think we should follow the way it's handled by psalm
https://psalm.dev/r/16085bfd30

**Comment by ondrejmirtes** [comment]
```php
// $s is string
if (class_exists($s)) {
    // should $s suddenly be a literal-string?
}
```

**Comment by craigfrancis** [comment]
[I’m out atm, and using phone, will try to check properly later]

The source of the variable has to be a literal, so if you got it from `example::class`, that’s fine… but if you’re checking the value contains a string that happens to match a class name (e.g. ‘class_exists()’), that doesn’t flip the variable to being a literal.

**Comment by ondrejmirtes** [comment]
That's what I thought, so this PR has to be fixed with my suggestion. Thank you.

**Comment by staabm** [comment]
thanks for the input guys. I just committed an update.

**Comment by staabm** [comment]
The spurious out of memory error could also be observed on other PRs. I don‘t think it is related to the actual change

File: src/Reflection/InitializerExprTypeResolver.php [hunk 1]
```diff
@@ -1726,11 +1726,17 @@ static function (Type $type, callable $traverse): Type {
 					if ($type instanceof TemplateType && !$type instanceof TypeWithClassName) {
 						return new GenericClassStringType($type);
 					} elseif ($type instanceof TypeWithClassName) {
+						$stringType = new GenericClassStringType($type);
+
 						$reflection = $type->getClassReflection();
 						if ($reflection !== null && $reflection->isFinalByKeyword()) {
-							return new ConstantStringType($reflection->getName(), true);
+							$stringType = new ConstantStringType($reflection->getName(), true);
```
Comment (rvanvelzen) [inline src/Reflection/InitializerExprTypeResolver.php]: `ConstantStringType` is already considered to be a literal string, so intersecting with `AccessoryLiteralStringType` will just remove the `AccessoryLiteralStringType` again.

Comment (staabm) [reply src/Reflection/InitializerExprTypeResolver.php]: great catch! fixed

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
I'd say that the ifs above should also be covered by this PR. Basically any `$obj::class` is a literal-string.

So with `$obj::class` and `$obj` being EnumCaseObjectType, it should also be literal-string. And with `$type instanceof TemplateType && !$type instanceof TypeWithClassName` that should work too.

**Comment by staabm** [comment]
adjusted per feedback.

I am thinking whether a `object`  typed parameter on which `::class` is appended should also be considered literal?

```php
/**
 * @param object $o
 */
function a($o): void
{
	assertType('class-string&literal-string', $o::class);
	sayHello($o::class);
}
```

**Comment by ondrejmirtes** [comment]
Of course :)

**Comment by ondrejmirtes** [comment]
Please rebase and solve the conflict, I'll merge it then.

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**